waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -0,0 +1,168 @@
+# Waku Noise Protocols for Waku Payload Encryption
+## See spec for more details:
+## https://github.com/vacp2p/rfc/tree/master/content/docs/rfcs/35
+##
+## Implementation partially inspired by noise-libp2p:
+## https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/secure/noise.nim
+
+{.push raises: [Defect].}
+
+import std/[oids, options, tables]
+import chronos
+import chronicles
+import bearssl
+import strutils
+import stew/[endians2]
+import nimcrypto/[utils, sha2, hmac]
+
+import libp2p/stream/[connection]
+import libp2p/peerid
+import libp2p/peerinfo
+import libp2p/protobuf/minprotobuf
+import libp2p/utility
+import libp2p/errors
+import libp2p/crypto/[crypto, chacha20poly1305, curve25519]
+
+
+when defined(libp2p_dump):
+  import libp2p/debugutils
+
+logScope:
+  topics = "nim-waku noise"
+
+const
+  # Empty is a special value which indicates k has not yet been initialized.
+  EmptyKey = default(ChaChaPolyKey)
+  NonceMax = uint64.high - 1 # max is reserved
+  NoiseSize = 32
+  MaxPlainSize = int(uint16.high - NoiseSize - ChaChaPolyTag.len)
+
+
+type
+  KeyPair* = object
+    privateKey: Curve25519Key
+    publicKey: Curve25519Key
+
+  NoisePublicKey* = object
+    flag: uint8
```

Comment:
What does `flag` indicate? would be good to add a comment

Reply:
Added comment in https://github.com/status-im/nim-waku/commit/704f3fe40fb4ad63212ec5f6face687b4971c909